torch/fx/subgraph_rewriter.py

```diff
@@ -265,15 +283,33 @@ def mark_node_as_replaced(n : Node) -> None:
         with original_graph.inserting_before(subgraph_output):
             copied_output = original_graph.graph_copy(replacement_graph,
                                                       val_map)
-        assert isinstance(copied_output, Node)
 
-        # We only want to copy in the output node from `pattern` if we
-        # have an output-output match. Otherwise, we leave out the
-        # `pattern` output node so we don't have two outputs in the
-        # resultant graph
+        # Hook the output Node of the replacement subgraph in to the
+        # original Graph at the correct location
+
+        # CASE 1: We need to hook the replacement subgraph in somewhere
+        # in the middle of the graph. We replace the Node in the
+        # original graph that corresponds to the end of the pattern
+        # subgraph
         if subgraph_output.op != "output":
-            subgraph_output = subgraph_output.args[0]    # type: ignore
-        subgraph_output.replace_all_uses_with(copied_output)
+            # `subgraph_output` may have multiple args. The arg that
+            # needs to be replaced by `copied_output` must have been
+            # a node that was originally matched as part of `pattern`.
+            for n in subgraph_output.args:
+                if n.op != "placeholder" and n in match.nodes_map.values():
+                    subgraph_output = n
+                    break
+            assert subgraph_output.op != "output"
+            subgraph_output.replace_all_uses_with(copied_output)
+        # CASE 2: The pattern subgraph match extends to the end of the
+        # original graph, so we need to change the current graph's
+        # output Node to reflect the insertion of the replacement graph
```

Comment:
I don't really understand what's going on here. The comment hints that we're changing something, but it doesn't say what or why, and when I read the code it's not obvious